src/general.c

```diff
@@ -242,8 +242,15 @@ static int do_create(const char *path, int nbranch_ro, int nbranch_rw) {
 
 	res = mkdir(dirp, buf.st_mode);
 	if (res == -1) {
-		USYSLOG(LOG_ERR, "Creating %s failed: \n", dirp);
-		RETURN(1);
+		if (errno == EEXIST) {
+			// In an NFS environment with many clients trying to write to the same directory tree
+			// and if that tree does not exist on the read write mount, there's a race between them.
+			// The directory may have been created by another client. It's not a fatal error.
+			USYSLOG(LOG_INFO, "Directory %s already existed - probably another client made it if using NFS", dirp);
+		} else {
+			USYSLOG(LOG_ERR, "Creating %s failed: \n", dirp);
+			RETURN(1);
+		}
 	}
```

Comment:
This part is ok

Reply:
setfile() below should executed by the thread that created the directory. 

Btw, this is not only an NFS issue, but could happen locally as well. At least once the VFS does not take and exclusive directory lock for such operations anymore. https://lkml.org/lkml/2022/6/15/1449

Reply:
In my use case only allowing the client which created the directory to chown then chmod the directory could cause problems

The most obvious case I can think of is when a directory has the set-group-ID bit set.

If the client which got the EEXIST error then goes on to create a file in the directory, and the client which is setting the user and group of the directory has not finished the chown and chmod, the new file will not have the correct group because at file create time the user on the directory will be incorrect.

In my situation I have 30ms roundtrip times into my NFS servers which is a lot of time for this to happen.

I think allowing all clients to run setfile() is probably the safest thing - even though there are still small windows where something might go wrong if the server resets the directory mode bits on a chown.

I don't think it's practical to try to solve this with locking. The problem with NFS (particularly v3) is locking is slow and usually people don't enable it because of the extra ports you have to open on firewalls etc, and administrative problems when a client dies and leaves a lock on the NFS server. 

Would it be acceptable if I made a PR which added an option to control if setfile is done when the directory is found to exist?

Reply:
Hmm yeah, right. In the ideal world this directory would be created as hidden file and only only renamed to the right file once meta operations are complete. And more ideal would direct access to the underlying file system and to create the dentry after the inode only.

Given that this is a special race condition, I wouldn't be opposed to the setfile() call, but it would be good to have comment to explain your intent as commit message and also in the code.

Reply:
how does this look?

https://github.com/rpodgorny/unionfs-fuse/compare/master...bryder:unionfs-fuse:add_nfs_tweak_option

I have yet to test it in production so it's not ready for a full PR but some feedback is appreciated before I do the final polishing of it.

I can see already I have some indent problems in there because I usually don't use tabs so I'll fix those.

Apart from formatting is this acceptable?

The reality is it's hard to do this kind of thing over NFS - it depends on the nfs version, the nfs server, the network etc etc. So it's more about reducing the errors.

Reply:
hmmm, ok... ...so, i have to leave for a week or so - please test your version in your production environment and if everything works, i'll merge it as soon as i return...